proposals/NNNN-isolated-any-functions.md

```diff
@@ -0,0 +1,516 @@
+the distributed actor as an `(any Actor)?` using the same mechanism
+as [`#isolation`](https://github.com/apple/swift-evolution/blob/main/proposals/0420-inheritance-of-actor-isolation.md#isolated-distributed-actors).
+
+### Isolation checking
+
+Since the isolation of a dynamically-isolated function value is
+statically unknown, calls to it typically cross an isolation barrier.
+This means that the call must be `await`ed even if the function is
+synchronous, and the arguments and result must satisfy the usual
+sendability restrictions for cross-isolation calls.
+
+If `f` is an immutable binding (such as a local `let` or a non-`inout`
+parameter) of `@isolated(any)` function type, then a call to `f` does
+not cross an isolation barrier if the current context is isolated
+to a derivation of the expression `f.isolation`.
+
+A context is isolated to a derivation of a expression `E` if it has an
+isolated capture that is initialized to a derivation of `E`.
+
+An expression is a derivation of `E` if:
+
+- it has the exact form of `E`;
+- it is a reference to a capture or immutable binding immediately
+  initialized with a derivation of `E`; or
+- it is a non-optional derivation of `E`.
+
+When analyzing an expression for these rules, certain non-instrumental
+differences in expression syntax and behavior must be ignored.  This
+and certain other terms in this section are intended to match the
+definitions laid out in [SE-0420](https://github.com/apple/swift-evolution/blob/main/proposals/0420-inheritance-of-actor-isolation.md#generalized-isolation-checking).
+
+For example:
+
+```swift
+func delay(operation: @isolated(any) () -> ()) {
+  let isolation = operation.isolation
+  Task { [isolated isolation] in
+    print("waking")
+    operation() // <-- does not cross an isolation barrier and so is synchronous
+    print("finished")
+  }
+}
+```
+
+TODO: Add some way to use `assumeIsolated` to assert that we're currently
+isolated to the isolation of an `@isolated(any)` function, such that
+calls within that context no longer cross isolation boundaries.
+Maybe this could also decompose the function value?
+
+### Adoption in task-creation routines
+
+There are a large number of functions in the standard library that create
+tasks:
+- `Task.init`
+- `Task.detached`
+- `TaskGroup.add`
+- `ThrowingTaskGroup.add`
+- `DiscardingTaskGroup.add`
+- `ThrowingDiscardingTaskGroup.add`
+
+This proposal modifies all of these APIs so that the task function has
+`@isolated(any)` function type.  These APIs now all synchronously enqueue
+the new task directly on the appropriate executor for the task function's
+dynamic isolation.
+
+Swift reserves the right to optimize the execution of tasks to avoid
+"unnecessary" isolation changes, such as when an isolated `async` function
+starts by calling a function with different isolation.  In general, this
+includes optimizing where the task initially starts executing.  As an
```

Comment:
This is probably worthy of an example as that's the "we'll reorder things" case that people should watch out for.

Maybe a bit here?

```suggestion
includes optimizing where the task initially starts executing:

```swift
@MainActor class Helper { 
  let other: OtherActor
  
  func help() {
    Task { await other.work() } // *implicitly* starts on MainActor, however 
    Task { await other.work() } // this may be optimized away and start on `other` immediately
  }
\``` // FIXME: can't suggest change with a \``` inside it, remove the \

In this example, `Task` is implicitly inheriting the enclosing context's MainActor isolation, however the task immediately calls code on another actor. The Swift runtime reserves the right to optimize away the initial "starting the task" enqueue on the MainActor, and instead enqueue on the `other` immediately.

As an
```

Did I get the semantics right of what you're proposing here?